### ISSUE TITLE
Loading the state dict of a model without InPlaceABN

### ISSUE THREAD

**Issue by yoniaflalo** [mapillary/inplace_abn#163]
Hi. 

First of all, this package is amazing, and provided an incredible speed boost to training all of my models, so thanks.

Now, I am trying to convert a standard pretrained model (like resnet50) to InPlaceABN model. Basically, I would like to be able to take any model, and apply on it a function that would convert all the BatchNorm2d with a InPlaceABN and copy all of the parameters.

I have written a the following script. It is working if I use ABN but it is not working when I use InPlaceABN.

Basically, I load a pretrain ResNet50 model, and apply the function `to_abn` on it. If I replace every `nn.BatchNorm2d` with `ABN`, I can infer on my test set and get the test accuracy I am supposed to.
If I use `InPlaceABN` in the below code, I get 0% accuracy, as if `InPlaceABN` was not able to load a state_dict from a regular `nn.BatchNorm2d`. Any idea where it comes from?

```
from inplace_abn import InPlaceABN, ABN
import torch.nn as nn
from utils import set_layer



def to_abn(module):
    if hasattr(module, 'module'):
        module = module.module
    for n, m in module.named_modules():
        if isinstance(m, nn.BatchNorm2d):
            num_features = m.num_features
            momentum = m.momentum
            eps = m.eps
            # The below line does not seems to work when I try to load the state_dict
            new_bn = InPlaceABN(num_features=num_features, eps=eps, momentum=momentum, activation='identity')
            # But the below line would work and provide same results as original network. 
            # new_bn = ABN(num_features=num_features, eps=eps, momentum=momentum, activation='identity')
            new_bn.load_state_dict(m.state_dict())
            set_layer(module, n, new_bn)
    return module

```

the function `set_layer` simply replaces a module of the model with another module.

Thanks for your help

**Comment by yoniaflalo** [comment]
I just found that the problem occurs in resnet when I try to convert the bn3, before the skip connection with identity or downsample. I do not understand this behavior.

The following piece of code would work:

```
import torchvision.models as models
from inplace_abn import InPlaceABN

def to_abn(module):
    if hasattr(module, 'module'):
        module = module.module
    for n, m in module.named_modules():
        if isinstance(m, nn.BatchNorm2d) and not 'bn3' in n:
            print(n)
            num_features = m.num_features
            momentum = m.momentum
            eps = m.eps
            new_bn = InPlaceABN(num_features=num_features, eps=eps, momentum=momentum, activation='identity')
            new_bn.load_state_dict(m.state_dict())
            set_layer(module, n, new_bn)
    return module

module = models.resnet50(pretrained=True)
module_abn = to_abn(module)
...
```

Then, the question is what happens with the third batch norm that prevent it to work? Maybe the fact that the operation is inplace can not handle the skip connection operation that is also inplace?

**Comment by ducksoup** [comment]
@yoniaflalo when you say that it doesn't work, are you talking about a crash, or is it just bad performance?

**Comment by ducksoup** [comment]
In any case, there are two issues with blindly replacing standard BN with InPlaceABN in the default ResNet50 model from Pytorch (or at least, this was the case last time I worked with it 2 years ago):
1) The default residual block implementation overwrites the output of the main branch with the sum between it and the skip connection. This conflicts with InPlaceABN and generally results in a crash during back-propagation.
2) InPlaceABN uses a different implementation of the scaling operation compared to standard BN and ABN, i.e. `y = (|w| + eps) * x + b` instead of `y = w * x + b`. This is because we need to ensure that the scaling is always invertible. Because of this, in general it is not possible to load pre-trained wegiths from a standard BN network in an InPlaceABN network.

I hope this helps.

**Comment by yoniaflalo** [comment]
I mean that the output is not equivalent to the regular batch norm. But I just understood why. Because in the module InplaceABN, you replace gamma with abs(gamma)+ epsilon. So the `nn.BatchNorm2d` is not mathematically equivalent to your InplaceABN. So I cannot load in InplaceABN a batchnorm state_dict that have been trained with  `nn.BatchNorm2d` .

**Comment by yoniaflalo** [comment]
Thanks a lot for your feedback. Now everything is clear. I believe I can close this issue since it is solved

**Comment by ducksoup** [comment]
@yoniaflalo there is actually a work around for this issue, even if a bit laborious. Assuming that every InPlaceABN/BN of the network is preceded by a standard Convolution with no bias and nothing else, you can adapt the weights of the pre-trained model to work correctly with InPlaceABN as follows:

For each Conv-BN pair `(conv, bn)` in the network do:
1) Find the negative bn weights: `mask = bn.weight < 0`
2) Invert the corresponding convolutional weights: `conv.weight[mask, :, :, :] *= -1`
3) Invert the bn weights: `bn.weight[mask] *= -1`

This should result in an (approximately because of `eps`) compatible model. Note that the procedure only works if the network exactly satisfies the condition I mentioned above.

**Comment by yoniaflalo** [comment]
Thanks a lot it's exactly what I was looking for

**Comment by Julymycin** [comment]
Hello, @yoniaflalo and @ducksoup , I now meet a seemly similar problem, I want to load a state_dict from InPlaceABN to BN because I want to deploy a model to C++  with libtorch.jit, while InPalceABN is not supported by JIT yet.
Looking through your discussion, it seems impossible to finish this work,  because all the InPlaceABN.weight >= 0,  I cannot decide which should be invert, am I right? Or is there a method to solve this problem?

**Comment by yoniaflalo** [comment]
> Hello, @yoniaflalo and @ducksoup , I now meet a seemly similar problem, I want to load a state_dict from InPlaceABN to BN because I want to deploy a model to C++ with libtorch.jit, while InPalceABN is not supported by JIT yet.
> Looking through your discussion, it seems impossible to finish this work, because all the InPlaceABN.weight >= 0, I cannot decide which should be invert, am I right? Or is there a method to solve this problem?

Hi @Julymycin . You have two options. The first one is the solution given before by @ducksoup , that works pretty good.

The second one is to perform Batch Norm Fusion. After having used ABN to train your network, you can remove all the batch norm by fusing it with the convolutional weights for deployment purpose. Here is a piece of code that can do this.

```
def fuse_bn_to_conv(bn_layer, conv_layer):
    # print('bn fuse')
    bn_st_dict = bn_layer.state_dict()
    conv_st_dict = conv_layer.state_dict()

    # BatchNorm params
    eps = bn_layer.eps
    mu = bn_st_dict['running_mean']
    var = bn_st_dict['running_var']
    gamma = bn_st_dict['weight']

    if 'bias' in bn_st_dict:
        beta = bn_st_dict['bias']
    else:
        beta = torch.zeros(gamma.size(0)).float().to(gamma.device)

    # Conv params
    W = conv_st_dict['weight']
    if 'bias' in conv_st_dict:
        bias = conv_st_dict['bias']
    else:
        bias = torch.zeros(W.size(0)).float().to(gamma.device)

    denom = torch.sqrt(var + eps)
    b = beta - gamma.mul(mu).div(denom)
    A = gamma.div(denom)
    bias *= A
    A = A.expand_as(W.transpose(0, -1)).transpose(0, -1)

    W.mul_(A)
    bias.add_(b)

    conv_layer.weight.data.copy_(W)
    if conv_layer.bias is None:
        conv_layer.bias = torch.nn.Parameter(bias)
    else:
        conv_layer.bias.data.copy_(bias)


def extract_layers(model):
    list_layers = []
    for n, p in model.named_modules():
        list_layers.append(n)
    return list_layers


def compute_next_bn(layer_name, resnet):
    list_layer = extract_layers(resnet)
    assert layer_name in list_layer
    if layer_name == list_layer[-1]:
        return None
    next_bn = list_layer[list_layer.index(layer_name) + 1]
    if extract_layer(resnet, next_bn).__class__.__name__ == 'BatchNorm2d':
        return next_bn
    return None


def fuse_bn(model):
    for n, m in model.named_modules():
        if isinstance(m, nn.Conv2d):
            next_bn = compute_next_bn(n, model)
            if next_bn is not None:
                next_bn_ = extract_layer(model, next_bn)
                fuse_bn_to_conv(next_bn_, m)
                set_layer(model, next_bn, nn.Identity())
    return model
```

You take your model and compute `fuse_bn(model)`. It will fuse every batch norm with the convolution and replace the batch norm with identity. (This piece of code was written for regular batch norm `nn.BatchNorm2d` but you can easily adapt it to `InplaceAbn`

**Comment by Julymycin** [comment]
Hi, @yoniaflalo , thanks for your help and sry for my late reply... I now understand the two methods, but how about the definition of `set_layer` and `exract_layer` in pytorch?  thx

**Comment by yoniaflalo** [comment]
Sure, I forgot, sorry.

```
def extract_layer(model, layer):
    layer = layer.split('.')
    module = model
    if hasattr(model, 'module') and layer[0] != 'module':
        module = model.module
    if not hasattr(model, 'module') and layer[0] == 'module':
        layer = layer[1:]
    for l in layer:
        if hasattr(module, l):
            if not l.isdigit():
                module = getattr(module, l)
            else:
                module = module[int(l)]
        else:
            return module
    return module


def set_layer(model, layer, val):
    layer = layer.split('.')
    module = model
    if hasattr(model, 'module') and layer[0] != 'module':
        module = model.module
    lst_index = 0
    module2 = module
    for l in layer:
        if hasattr(module2, l):
            if not l.isdigit():
                module2 = getattr(module2, l)
            else:
                module2 = module2[int(l)]
            lst_index += 1
    lst_index -= 1
    for l in layer[:lst_index]:
        if not l.isdigit():
            module = getattr(module, l)
        else:
            module = module[int(l)]
    l = layer[lst_index]
    setattr(module, l, val)
```

**Comment by Julymycin** [comment]
Thank you @yoniaflalo ! I deploy successfully by using your methods. Looking through some issues in this repo, I also try to replace InplaceABNSync with ABN by change the` *.bn*.weight` to `abs(*.bn*.weight)+eps`, because in the network, the activation parameter of some InPlaceABNSync are set to none, while some are set to default. Even though it seems to cost 80% more GPU memory, the precision is great. Thanks!

**Comment by poornimajd** [comment]
@Julymycin Could you please let me know where exactly should we change this ?

> ABN by change the` *.bn*.weight` to `abs(*.bn*.weight)+eps`

Is it when I load the model and then look into each layer and try to do this?Or is it in the definition of the model?

**Comment by Julymycin** [comment]
@poornimajd here is a sample
`model = networks.init_model('resnet101', num_classes=num_classes, pretrained=None)


    state_dict = torch.load(**inplaceABN_model_path**)['state_dict']
    from collections import OrderedDict
    new_state_dict = OrderedDict()
    for k, v in state_dict.items():
        k = k[7:]
        ss = k.split('.')
        if ss[-2].startswith('bn') and ss[-1].endswith('weight'):
            v1 = torch.abs(v) + eps
        else:
            v1 = v
        new_state_dict[k] = v1
    model.load_state_dict(new_state_dict)
    model.cuda()
    model.eval()
    torch.save(model.state_dict(), **ABN_model_path**)`

**Comment by LOOKCC** [comment]
@Julymycin  I want to ask you that is there any way to change Inplace-abn to BN for training? The method above is  for deployment.

**Comment by Chen-Song** [comment]
> model = networks.init_model('resnet101', num_classes=num_classes, pretrained=None)

Hi, in 'model = networks.init_model('resnet101', num_classes=num_classes, pretrained=None)', it use standard BN or Inplace_ABN?
